### PR TITLE
chore(ci_visibility): add quarantine support to pytest

### PR DESCRIPTION
**Pull request by vitor-de-araujo**
This PR adds preliminary support for quarantining tests in pytest.

The API to query which tests are quarantined does not exist yet on the backend side, and the final form of that API is still to be defined, so the code dealing with the API has been moved to a separate PR (https://github.com/DataDog/dd-trace-py/pull/11770). Currently, we can mark tests as quarantined by manually adding the `test.quarantine.is_quarantined` tag to the test with a pytest decorator:

```
@pytest.mark.dd_tags(**{"test.quarantine.is_quarantined": True})
def test_fail_quarantined():
    assert False
```

For testing purposes, the environment variables `_DD_TEST_FORCE_ENABLE_QUARANTINE` and `_DD_TEST_FORCE_ENABLE_ATR` have been added to enable quarantine and ATR without depending on the backend.

The test reporting looks like below. Errors and logs for quarantined tests are not printed.
image

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/contrib/pytest/_atr_utils.py [hunk 1]
```diff
@@ -29,41 +29,69 @@ class _ATR_RETRY_OUTCOMES:
...
     if when == "call":
         if test_outcome.status == TestStatus.FAIL:
-            original_result.outcome = _ATR_RETRY_OUTCOMES.ATR_ATTEMPT_FAILED
-        return
-    if InternalTest.get_tag(test_id, "_dd.ci.atr_setup_failed"):
-        log.debug("Test item %s failed during setup, will not be retried for Early Flake Detection")
-        return
-    if InternalTest.get_tag(test_id, "_dd.ci.atr_teardown_failed"):
-        # NOTE: tests that passed their call but failed during teardown are not retried
-        log.debug("Test item %s failed during teardown, will not be retried for Early Flake Detection")
```
Comment (vitor-de-araujo) [inline ddtrace/contrib/pytest/_atr_utils.py:55]: This was a relic from a previous incarnation of the code, these tags are not used anymore; instead this is handled in [_pytest_runtest_makereport](https://github.com/DataDog/dd-trace-py/blob/vitor-de-araujo%2FSDTEST-1255%2Fquarantine/ddtrace/contrib/pytest/_plugin_v2.py#L452) through stashed values.

File: ddtrace/internal/ci_visibility/_api_client.py [hunk 2]
```diff
@@ -372,6 +381,12 @@ def fetch_settings(self) -> TestVisibilityAPISettings:
                 )
             else:
                 early_flake_detection = EarlyFlakeDetectionSettings()
+
+            quarantine = QuarantineSettings(
+                enabled=attributes.get("quarantine", {}).get("enabled", False)
+                or asbool(os.getenv("_DD_TEST_FORCE_ENABLE_QUARANTINE"))  # ꙮ
```
Comment (vitor-de-araujo) [inline ddtrace/internal/ci_visibility/_api_client.py:387]: This is here to make my life easier to test quarantine while we don't have the backend, we don't need to keep it (or we can move it somewhere else).

File: ddtrace/internal/ci_visibility/_api_client.py [hunk 3]
```diff
@@ -525,6 +544,77 @@ def fetch_unique_tests(self) -> t.Optional[t.Set[InternalTestId]]:
 
         return unique_test_ids
 
+    def fetch_test_details(self):
+        metric_names = APIRequestMetricNames(
+            count=EARLY_FLAKE_DETECTION_TELEMETRY.REQUEST,
+            duration=EARLY_FLAKE_DETECTION_TELEMETRY.REQUEST_MS,
+            response_bytes=EARLY_FLAKE_DETECTION_TELEMETRY.RESPONSE_BYTES,
+            error=EARLY_FLAKE_DETECTION_TELEMETRY.REQUEST_ERRORS,
+        )  # ꙮ
```
Comment (vitor-de-araujo) [inline ddtrace/internal/ci_visibility/_api_client.py:553]: I need to replace these with new metrics.

**Review by romainkomorndatadog — APPROVED** [review]

File: tests/contrib/pytest/test_pytest_quarantine.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+            return_value=TestVisibilityAPISettings(
+                quarantine=QuarantineSettings(enabled=True),
+                flaky_test_retries_enabled=False,
+            ),
+        ):
+            yield
+            from ddtrace.internal.ci_visibility.recorder import CIVisibility
+
+            if CIVisibility.enabled:
+                CIVisibility.disable()
```
[the remark this answers is not in the document]
Comment (vitor-de-araujo) [reply tests/contrib/pytest/test_pytest_quarantine.py:112]: I copied it from https://github.com/DataDog/dd-trace-py/blob/vitor-de-araujo%2FSDTEST-1255%2Fquarantine/tests/contrib/pytest/test_pytest_efd.py#L121, but maybe it's unnecessary there as well.

[the remark this answers is not in the document]
Comment (romainkomorndatadog) [reply tests/contrib/pytest/test_pytest_quarantine.py:112]: Oh. Whoever wrote probably wasn't paying attention... Let's see what `git blame` says... 😬

[the remark this answers is not in the document]
Comment (romainkomorndatadog) [reply tests/contrib/pytest/test_pytest_quarantine.py:112]: (but actually, I think I added it to the `inline_run` method after I wrote the EFD tests. Maybe.)

[the remark this answers is not in the document]
Comment (vitor-de-araujo) [reply tests/contrib/pytest/test_pytest_quarantine.py]: I've removed it from all the classes, let's see how that goes https://github.com/DataDog/dd-trace-py/pull/11615/commits/1ac8026fb1342dfe4da0e59d7da5d46f6c9d606a

**Pull request merged**